jobsdb/backup.go

```diff
@@ -51,37 +51,68 @@ func (jd *HandleT) backupDSLoop(ctx context.Context) {
 		case <-ctx.Done():
 			return
 		}
-		jd.logger.Debugf("backupDSLoop backup enabled %s", jd.tablePrefix)
-		backupDSRange := jd.getBackupDSRange()
-		// check if non-empty dataset is present to back up
-		// else continue
-		sleepMultiplier = 1
-		if (dataSetRangeT{} == *backupDSRange) {
-			// sleep for more duration if no dataset is found
-			sleepMultiplier = 6
-			continue
-		}
+		loop := func() error {
+			jd.logger.Debugf("backupDSLoop backup enabled %s", jd.tablePrefix)
+			backupDSRange, err := jd.getBackupDSRange()
+			if err != nil {
+				return fmt.Errorf("[JobsDB] :: Failed to get backup dataset range. Err: %w", err)
+			}
+			// check if non-empty dataset is present to back up
+			// else continue
+			sleepMultiplier = 1
+			if (dataSetRangeT{} == *backupDSRange) {
+				// sleep for more duration if no dataset is found
+				sleepMultiplier = 6
+				return nil
+			}
 
-		backupDS := backupDSRange.ds
+			backupDS := backupDSRange.ds
 
-		opPayload, err := json.Marshal(&backupDS)
-		jd.assertError(err)
+			opPayload, err := json.Marshal(&backupDS)
+			jd.assertError(err)
 
-		opID := jd.JournalMarkStart(backupDSOperation, opPayload)
-		err = jd.backupDS(ctx, backupDSRange)
-		if err != nil {
-			jd.logger.Errorf("[JobsDB] :: Failed to backup jobs table %v. Err: %v", backupDSRange.ds.JobStatusTable, err)
+			opID, err := jd.JournalMarkStart(backupDSOperation, opPayload)
+			if err != nil {
+				return fmt.Errorf("mark start of backup operation: %w", err)
+			}
+			err = jd.backupDS(ctx, backupDSRange)
+			if err != nil {
+				return fmt.Errorf("backup dataset: %w", err)
+			}
+			err = jd.JournalMarkDone(opID)
+			if err != nil {
+				return fmt.Errorf("mark end of backup operation: %w", err)
+			}
+
+			// drop dataset after successfully uploading both jobs and jobs_status to s3
+			opID, err = jd.JournalMarkStart(backupDropDSOperation, opPayload)
+			if err != nil {
+				return fmt.Errorf("mark start of drop backup operation: %w", err)
+			}
+			// Currently, we retry uploading a table for some time & if it fails. We only drop that table & not all `pre_drop` tables.
+			// So, in situation when new table creation rate is more than drop. We will still have pipe up issue.
+			// An easy way to fix this is, if at any point of time exponential retry fails then instead of just dropping that particular
+			// table drop all subsequent `pre_drop` table. As, most likely the upload of rest of the table will also fail with the same error.
+			err = jd.dropDS(backupDS)
+			if err != nil {
+				return fmt.Errorf(" drop dataset: %w", err)
+			}
+			err = jd.JournalMarkDone(opID)
+			if err != nil {
+				return fmt.Errorf("mark end of drop backup operation: %w", err)
+			}
+			return nil
+		}
+		if err := loop(); err != nil {
+			if !jd.skipMaintenanceError {
+				if ctx.Err() == nil {
+					panic(err)
+				}
+				return
+			}
+			jd.logger.Errorf("[JobsDB] :: Failed to backup dataset. Err: %s", err.Error())
 		}
```

Comment:
We shouldn't log an error if the parent context is cancelled. In this scenario, the function will return immediately after the next for loop.
```suggestion
		if err := loop(); err != nil && ctx.Err() == nil {
			if !jd.skipMaintenanceError {
				panic(err)
			}
			jd.logger.Errorf("[JobsDB] :: Failed to backup dataset. Err: %s", err.Error())
		}
```